content/4_Gold/Digit_DP.mdx

```diff
@@ -92,61 +115,276 @@ void reset() {
+	public static void reset() {
+		for (int i = 0; i < 19; i++) {
+			for (int j = 0; j < 50; j++) {
+				for (int k = 0; k < 2; k++) {
+					for (int l = 0; l < 2; l++) { dp[i][j][k][l] = -1; }
+				}
+			}
+		}
+	}
+
+	/**
+	 * Calculate the number of sequences whose digits consist of at least half
+	 * targ. If targ2 is not -1, then calculate the number of sequences whose
+	 * digits are exactly half targ and half targ2.
+	 * @param pos     the starting position
+	 * @param k       counter for the number of the target digit targ
+	 * @param under   1 if the current sequence is smaller than the given
+	 *     upperbound
+	 * @param started 1 if the sequence has any digit other than leading zeros
+	 */
+	public static long solveDP(int pos, int k, int under, int started, int targ,
+	                           int targ2) {
+		// base case: the sequence has reached the maximal length
+		if (pos == num.length()) {
+			// edge case: the sequence only contains zeros, i.e. did not start
+			if (started == 0) { return 0; }
+
+			/*
+			 * If we are interested in finding out whether the sequence contains
+			 * exactly half digits of targ and half digits of targ2.
+			 */
+			if (targ2 != -1) {
+				if (k == 20) {
+					return 1;
+				} else {
+					return 0;
+				}
+			}
+			/*
+			 * Otherwise, the sequence is interesting if at least half of the
+			 * digits are our target number targ.
+			 */
+			if (k >= 20) {
+				return 1;
+			} else {
+				return 0;
+			}
+		}
+
+		if (dp[pos][k][under][started] != -1) {
+			return dp[pos][k][under][started];
+		}
+
+		long ans = 0;
+		for (int i = 0; i <= 9; i++) {
+			int digit_diff = num.charAt(pos) - '0';
+			/*
+			 * if the sequence will be larger than the upperbound, then we can
+			 * terminate
+			 */
+			if (under == 0 && i > digit_diff) { break; }
+
+			/*
+			 * If the new digit is smaller than the one in the upperbound, then
+			 * any child sequence can't be larger than than upperbound as this
+			 * digit is more significant than any following one
+			 */
+			int isUnder = under;
+			if (i < digit_diff) { isUnder = 1; }
+
+			// the sequence has started if any digits until pos is other than 0
+			int isStarted = started | (i != 0 ? 1 : 0);
+
+			/*
+			 * If we want to count how many sequences have exactly half digits
+			 * of targ and half of targ2, then any other number than targ or
+			 * targ2 doesn't matter
+			 */
+			if (isStarted == 1 && targ2 != -1 && i != targ && i != targ2) {
+				continue;
+			}
+
+			/*
+			 * The count for target digit targ is increased by one if i is targ,
+			 * or decreased by one otherwise.
+			 */
+			int newK = k;
+			if (isStarted == 1) {
+				if (targ == i) {
+					newK = k + 1;
+				} else {
+					newK = k - 1;
+				}
+			}
+			ans += solveDP(pos + 1, newK, isUnder, isStarted, targ, targ2);
+		}
+		return dp[pos][k][under][started] = ans;
+	}
+
+	/** Count interesting sequences that are less or equal ubound */
```

Comment:
```suggestion
	// Count interesting sequences that are less or equal ubound
```

Reply:
emm we do need `**` for function doc don't we?

Reply:
yeah dw about this